### PR TITLE
prov/shm: add shm signal to reduce lock contention

### PR DESCRIPTION
**Pull request by aingerson**
In order to make progress, shm needs to check the cmd list for any
incoming messages and the resp queue for any operation completion
notifications. The only protection for these is the shm lock. This
can cause a lot of lock contention when the EP is trying to make
progress on incoming messages while a peer it trying to send a
message.

These issues are especially evident on particular nodes. Unclear
what the actual cause of the increase in contention is.

Regardless, to fix this, add a signal to the shm region which is used
to check whether there could be an incoming cmd or response. Peers can
still send messages without competing with the target for the lock.
Whenever a cmd is inserted or a response is updated, signal the lock
to let the EP know to try to make progress.

See performance impacts below on fi_rdm_pingpong.

Usual performance disclaimer: this is from single runs on one system;
this is not meant as an official performance snapshot.

Before (bad node):

```bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      10k     1.2m       18.46s      0.07     923.11       0.00
256     10k     4.8m       22.74s      0.23    1137.00       0.00
1k      10k     19m        20.03s      1.02    1001.65       0.00
4k      10k     78m        15.53s      5.28     776.40       0.00
64k     1k      125m        1.85s     70.86     924.84       0.00
1m      100     200m        0.25s    849.42    1234.46       0.00
```

Before (good node):
```bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      10k     1.2m        0.04s     29.00       2.21       0.45
256     10k     4.8m        0.05s    106.42       2.41       0.42
1k      10k     19m         0.05s    420.15       2.44       0.41
4k      10k     78m         0.07s   1109.17       3.69       0.27
64k     1k      125m        0.02s   8009.78       8.18       0.12
1m      100     200m        0.02s  10770.09      97.36       0.01
```

After (bad node):
```bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      10k     1.2m        0.03s     40.18       1.59       0.63
256     10k     4.8m        0.04s    143.14       1.79       0.56
1k      10k     19m         0.04s    562.14       1.82       0.55
4k      10k     78m         0.06s   1407.92       2.91       0.34
64k     1k      125m        0.02s   7380.18       8.88       0.11
1m      100     200m        0.02s   9395.42     111.61       0.01
```

After (good node):
```bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      10k     1.2m        0.02s     69.94       0.92       1.09
256     10k     4.8m        0.02s    247.39       1.03       0.97
1k      10k     19m         0.02s    939.49       1.09       0.92
4k      10k     78m         0.03s   2411.68       1.70       0.59
64k     1k      125m        0.01s  10310.08       6.36       0.16
1m      100     200m        0.02s  10885.81      96.32       0.01
```

The patch also seems to resolve issues when running shm in a non-debug build
which causes random performance issues. See performance impacts below:

Non-debug build before (sample random bad run, appeared after 2 runs):
```bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      10k     1.2m        1.59s      0.81      79.41       0.01
256     10k     4.8m        0.54s      9.51      26.91       0.04
1k      10k     19m         0.21s     99.54      10.29       0.10
4k      10k     78m         3.32s     24.68     165.98       0.01
64k     1k      125m        0.46s    284.46     230.38       0.00
1m      100     200m        0.04s   4968.26     211.05       0.00
```

Non-debug build after (consistent with 50 runs):

```bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      10k     1.2m        0.01s    120.63       0.53       1.88
256     10k     4.8m        0.01s    391.23       0.65       1.53
1k      10k     19m         0.01s   1425.69       0.72       1.39
4k      10k     78m         0.03s   3202.00       1.28       0.78
64k     1k      125m        0.01s  10509.30       6.24       0.16
1m      100     200m        0.02s  10098.00     103.84       0.01
```

Signed-off-by: aingerson <alexia.ingerson@intel.com>

### CODE REVIEW

**Comment by aingerson** [comment]
@shefty @nikhilnanal This patch should resolve the CI performance issues. It should also let us enable the shm non-debug testing.

Added do not merge label so I can go through the entire testing logs and do more thorough testing before merging. Will remove when I feel confident about all bases.

File: prov/shm/src/smr_progress.c
Comment (shefty) [inline prov/shm/src/smr_progress.c:1032]: It seems that this code should acquire the signal lock before reading it.  The value of signal also seems opposite from what I expected.  Signal is acting as a bool, but false indicates the signaled state.

Based on the changes, an atomic value here would be better.

Comment (aingerson) [reply prov/shm/src/smr_progress.c:1032]: Fair. Better?

Comment (shefty) [reply prov/shm/src/smr_progress.c:1032]: Better.  Best would be:
```
if (ofi_atomic_cas_bool32(&ep->region->signal, 1, 0)) {
    smr_progress_resp(ep);
    smr_progress_cmd(ep);
    smr_progress_sar_list(ep);
}
```

Comment (aingerson) [reply prov/shm/src/smr_progress.c:1032]: 👍 Updated

**Review by shefty — APPROVED** [review]

**Comment by aingerson** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
We will look into it and get back to you.

**Comment by aingerson** [comment]
@wzamazon Thanks!
@shefty Should be good to go unless you have any further comments

**Pull request merged**
